File: python/lib/exporter.py
"""Output formatting: console table, CSV, and JSON export."""

from __future__ import annotations

import csv
import json
import sys
from pathlib import Path
from typing import Sequence

from lib.detector import DetectionResult

# Column widths for console output
_COL_FILE = 50
_COL_ENCODING = 14
_COL_CONFIDENCE = 12
_COL_SIZE = 12


def _confidence_str(result: DetectionResult) -> str:
    """Format confidence as a percentage string or dash."""
    if result.error:
        return "ERROR"
    if result.confidence is None:
        return "—"
    return f"{result.confidence * 100:.0f}%"


def _size_str(size: int) -> str:
    """Human-readable file size."""
    if size < 1024:
        return f"{size} B"
    elif size < 1024 ** 2:
        return f"{size / 1024:.1f} KB"
    else:
        return f"{size / 1024 ** 2:.1f} MB"


def _truncate(text: str, width: int) -> str:
    """Truncate a string to fit within `width` characters."""
    if len(text) <= width:
        return text
    return "…" + text[-(width - 1):]


def print_console(results: Sequence[DetectionResult], root: Path) -> None:
    """Print results as a formatted table to stdout.

    Args:
        results: Sequence of DetectionResult objects.
        root: The scanned root folder, used to compute relative paths.
    """
    header = (
        f"{'File':<{_COL_FILE}}  "
        f"{'Encoding':<{_COL_ENCODING}}  "
        f"{'Confidence':<{_COL_CONFIDENCE}}  "
        f"{'Size':>{_COL_SIZE}}"
    )
    separator = "  ".join([
        "-" * _COL_FILE,
        "-" * _COL_ENCODING,
        "-" * _COL_CONFIDENCE,
        "-" * _COL_SIZE,
    ])

    print(header)
    print(separator)

    for r in results:
        try:
            rel = r.path.relative_to(root)
        except ValueError:
            rel = r.path

        file_col = _truncate(str(rel), _COL_FILE)
        enc_col = r.encoding or ("ERROR" if r.error else "unknown")
        conf_col = _confidence_str(r)
        size_col = _size_str(r.size_bytes)

        print(
            f"{file_col:<{_COL_FILE}}  "
            f"{enc_col:<{_COL_ENCODING}}  "
            f"{conf_col:<{_COL_CONFIDENCE}}  "
            f"{size_col:>{_COL_SIZE}}"
        )

    print(separator)
    errors = sum(1 for r in results if r.error)
    print(f"\n{len(results)} file(s) scanned. {errors} error(s).")


def export_csv(results: Sequence[DetectionResult], output_path: Path, root: Path) -> None:
    """Write results to a CSV file.

    Args:
        results: Sequence of DetectionResult objects.
        output_path: Destination file path.
        root: The scanned root folder, used to compute relative paths.
    """
    with output_path.open("w", newline="", encoding="utf-8") as f:
        writer = csv.DictWriter(
            f,
            fieldnames=["path", "encoding", "confidence", "size_bytes", "error"],
        )
        writer.writeheader()
        for r in results:
            try:
                rel = str(r.path.relative_to(root))
            except ValueError:
                rel = str(r.path)

            writer.writerow(
                {
                    "path": rel,
                    "encoding": r.encoding or "",
                    "confidence": f"{r.confidence:.4f}" if r.confidence is not None else "",
                    "size_bytes": r.size_bytes,
                    "error": r.error or "",
                }
            )
    print(f"CSV exported → {output_path}", file=sys.stderr)


def export_json(results: Sequence[DetectionResult], output_path: Path, root: Path) -> None:
    """Write results to a JSON file.

    Args:
        results: Sequence of DetectionResult objects.
        output_path: Destination file path.
        root: The scanned root folder, used to compute relative paths.
    """
    data = []
    for r in results:
        try:
            rel = str(r.path.relative_to(root))
        except ValueError:
            rel = str(r.path)

        data.append(
            {
                "path": rel,
                "encoding": r.encoding,
                "confidence": round(r.confidence, 4) if r.confidence is not None else None,
                "size_bytes": r.size_bytes,
                "error": r.error,
            }
        )

    with output_path.open("w", encoding="utf-8") as f:
        json.dump(data, f, indent=2, ensure_ascii=False)

    print(f"JSON exported → {output_path}", file=sys.stderr)

File: nodejs/lib/exporter.js
/**
 * @file exporter.js
 * Output formatting: console table, CSV, and JSON export.
 */

import { writeFile } from "fs/promises";
import { relative } from "path";

// Column widths for console output
const COL_FILE = 50;
const COL_ENC = 14;
const COL_CONF = 12;
const COL_SIZE = 12;

/**
 * Format a confidence score as a percentage string.
 * @param {import('./detector.js').DetectionResult} result
 * @returns {string}
 */
function fmtConfidence(result) {
  if (result.error) return "ERROR";
  if (result.confidence === null) return "—";
  return `${Math.round(result.confidence * 100)}%`;
}

/**
 * Format byte count as human-readable size.
 * @param {number} bytes
 * @returns {string}
 */
function fmtSize(bytes) {
  if (bytes < 1024) return `${bytes} B`;
  if (bytes < 1024 ** 2) return `${(bytes / 1024).toFixed(1)} KB`;
  return `${(bytes / 1024 ** 2).toFixed(1)} MB`;
}

/**
 * Truncate a string to fit within `width` characters, with a leading ellipsis.
 * @param {string} text
 * @param {number} width
 * @returns {string}
 */
function truncate(text, width) {
  if (text.length <= width) return text;
  return "\u2026" + text.slice(-(width - 1));
}

/**
 * Pad or truncate a string to an exact width (left-aligned).
 * @param {string} text
 * @param {number} width
 * @returns {string}
 */
function padEnd(text, width) {
  return text.slice(0, width).padEnd(width);
}

/**
 * Print detection results as a formatted table to stdout.
 *
 * @param {import('./detector.js').DetectionResult[]} results
 * @param {string} rootDir - Scanned root directory for relative paths.
 * @param {object} [chalk]  - Optional chalk instance for colour output.
 */
export function printConsole(results, rootDir, chalk) {
  const c = chalk;

  const header = [
    padEnd("File", COL_FILE),
    padEnd("Encoding", COL_ENC),
    padEnd("Confidence", COL_CONF),
    "Size".padStart(COL_SIZE),
  ].join("  ");

  const sep = [
    "-".repeat(COL_FILE),
    "-".repeat(COL_ENC),
    "-".repeat(COL_CONF),
    "-".repeat(COL_SIZE),
  ].join("  ");

  console.log(c ? c.bold(header) : header);
  console.log(sep);

  for (const r of results) {
    const rel = relative(rootDir, r.path) || r.path;
    const file = truncate(rel, COL_FILE);
    const enc = r.error ? "ERROR" : (r.encoding ?? "unknown");
    const conf = fmtConfidence(r);
    const size = fmtSize(r.sizeBytes);

    const line = [
      padEnd(file, COL_FILE),
      padEnd(enc, COL_ENC),
      padEnd(conf, COL_CONF),
      size.padStart(COL_SIZE),
    ].join("  ");

    if (r.error && c) {
      console.log(c.red(line));
    } else {
      console.log(line);
    }
  }

  const errorCount = results.filter((r) => r.error).length;
  console.log(sep);
  console.log("");
  const summary = `${results.length} file(s) scanned. ${errorCount} error(s).`;
  console.log(c ? c.dim(summary) : summary);
}

/**
 * Write results to a CSV file.
 *
 * @param {import('./detector.js').DetectionResult[]} results
 * @param {string} outputPath - Destination file path.
 * @param {string} rootDir    - Scanned root directory for relative paths.
 */
export async function exportCsv(results, outputPath, rootDir) {
  const header = "path,encoding,confidence,size_bytes,error";
  const rows = results.map((r) => {
    const rel = relative(rootDir, r.path) || r.path;
    const enc = r.encoding ?? "";
    const conf = r.confidence !== null ? r.confidence.toFixed(4) : "";
    const err = r.error ?? "";
    // Escape fields containing commas or quotes
    const esc = (v) => (/[",\n]/.test(v) ? `"${v.replace(/"/g, '""')}"` : v);
    return [esc(rel), esc(enc), conf, r.sizeBytes, esc(err)].join(",");
  });

  const csv = [header, ...rows].join("\n");
  await writeFile(outputPath, csv, "utf-8");
  process.stderr.write(`CSV exported -> ${outputPath}\n`);
}

/**
 * Write results to a JSON file.
 *
 * @param {import('./detector.js').DetectionResult[]} results
 * @param {string} outputPath - Destination file path.
 * @param {string} rootDir    - Scanned root directory for relative paths.
 */
export async function exportJson(results, outputPath, rootDir) {
  const data = results.map((r) => ({
    path: relative(rootDir, r.path) || r.path,
    encoding: r.encoding,
    confidence:
      r.confidence !== null ? parseFloat(r.confidence.toFixed(4)) : null,
    size_bytes: r.sizeBytes,
    error: r.error,
  }));

  await writeFile(outputPath, JSON.stringify(data, null, 2), "utf-8");
  process.stderr.write(`JSON exported -> ${outputPath}\n`);
}
